lib/manager/gradle-wrapper/artifacts.ts

```diff
@@ -80,11 +80,16 @@ export async function updateArtifacts({
       cmd += ` --gradle-distribution-url ${distributionUrl}`;
       if (newPackageFileContent.includes('distributionSha256Sum=')) {
         // need to reset version, otherwise we have a checksum mismatch
+        const checksum = await getDistributionChecksum(distributionUrl);
         await writeLocalFile(
           packageFileName,
-          newPackageFileContent.replace(config.newValue, config.currentValue)
+          newPackageFileContent
+            .replace(config.newValue, config.currentValue)
+            .replace(
+              /distributionSha256Sum=.*/,
+              `distributionSha256Sum=${checksum}`
+            )
         );
```

Comment:
How does the file look after this operation?

I'm a bit surprised here since it's actually doing the following:

- Replace the _new_ version with the _old_
- But set the _new_ checksum

I can't see how that can be consistent. Old version with new checksum doesn't make sense?

I think, as I noted in https://github.com/renovatebot/renovate/issues/13081#issuecomment-992302289, that `replace(config.newValue, config.currentValue)` is not actually doing anything. If that's true it should be OK to remove that operation, but the comment above this code should be changed.

Reply:
For the gradle specifics its worth reading https://docs.gradle.org/current/userguide/gradle_wrapper.html#sec:upgrading_wrapper

As it's noted the gradle wrapper should be run _twice_ to also update the other wrapper files. This is because the first time we run gradle it will run on the old version, and second time it will run using the new version, and only then update other files.

If we do _not_ reset version, we have actually done the first run using Renovate alone. So the "first" gradle run will act as the second run. It will download the _new_ version and upgrade the related wrapper files using the correct version.

Reply:
@henrist Here is how it'll look : https://github.com/hasanwhitesource/renovate-repro-gradle-wrapper/blob/9aad840f190804136b6ae918aaf281b8c5f80b12/gradle/wrapper/gradle-wrapper.properties#L3
`replace(config.newValue, config.currentValue)` is replacing the new version with the old version in your gradle-wrapper.properties.
It was updating the gradle with no problem but we weren't setting the new sha.

Reply:
You are referring to the final result, not the intermediate result I was refering to. I'm thinking of how it looks at the time the "gradlew" command is run (the data written by the call to `writeLocalFile`).

The code _tries_ to "revert" the contents of the file. But I suspect it is not doing that for some reason. Looking at the contents of the file _at this point in code_ will show if the replace operation worked or not.

Reply:
@henrist I have tried it and also saw that the command is being executed with the correct sha. The gradle command was executing correctly.
The problem here was simply we didn't have a logic that was updating the sha in the file (in case of a sha)

This is the "gradlew" command that is being executed in your case : 
`'gradlew.bat wrapper --gradle-distribution-url https://services.gradle.org/distributions/gradle-7.4-all.zip --gradle-distribution-sha256-sum cd5c2958a107ee7f0722004a12d0f8559b4564c34daad7df06cffd4d12a426d0'`

This is the correct sha.

Reply:
We previously reset the version in file to the old, because it was already done by the renovate auto replace.

If we then run Gradle without resetting, there is a checksum mismatch, as the new version and the old checksum collides.

Which checksum is written with this change?

Reply:
> Note that running the wrapper task once will update gradle-wrapper.properties only, but leave the wrapper itself in gradle-wrapper.jar untouched. This is usually fine as new versions of Gradle can be run even with ancient wrapper files. If you nevertheless want all the wrapper files to be completely up-to-date, you’ll need to run the wrapper task a second time.

So instead of reset the version to the old value, we need to update the checksum to the new value and then run Gradle to update wrapper.

The checksum should only be fetched, if there is an existing value.

Reply:
@viceice  @henrist The new checksum is being written as see.
Here is an example repo : https://github.com/hasanwhitesource/renovate-repro-gradle-wrapper/pull/12/files#diff-40640fe1078ece83d7ea8fb67daacd77923a86d13447baf9769660b3b46f2eceR3

Reply:
@viceice So we only need to update the sha in the file and remove the change from old version to new version?
Tried it that way and it worked, just want to make sure if updating the file is necessary here.

Reply:
It can be done here to fix the issue.

A better solution would be to use renovate digest update, so the checksum will be updated in same step as the new version is written.